proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+It's also easier to find documentation for this syntax compared to `'` and `''`.
+
+#### Punning usage warnings
+
+We also propose to introduce two new warnings to GHC: `-Wpuns` and
+`-Wpun-bindings`.
+
+* `-Wpuns` warns the user about the usage of punning at use sites.
+
+* `-Wpun-bindings` warns the user about the introduction of punning at binding
+  sites.
+
+These warnings aim to help the user to avoid using punning in their codebase.
+
+To determine whether some binding or use site takes advantage of punning we
+ask the question: "If Haskell had a single unified namespace, would that change
+the lookup result?". If the answer is yes, then the code uses punning:
+
+```haskell
+data T = T -- with a single unified namespace, T the data would have a name clash with T the type.
+
+a = 15
+
+f :: forall a. a -> a -- no warning, because with a single unified namespace 'forall a.' would shadow the top level 'a', so 'a' in 'a -> a' would still refer to the forall bound 'a'.
+```
+
+There are more detailed examples in Examples section.
+
+## Proposed Change Specification
+
+* To disambiguate the namespaces of identifiers from modules that use punning,
+  the `-XExplicitNamespaces` extension is extended with new syntax:
+
+  ```
+  impdecl   -> import [qualified] modid [as modid] [impspec]
+  ```
+  is changed to
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+
+* To disambiguate the namespaces within one module, a `-XModuleAliases` extension
+  is added that introduces new syntax:
+
+  ```
+  module   -> module modid [exports] where body
+            | body
+  ```
+  is changed to
+  ```
+  module   -> module modid [alias_1, ..., alias_n] [exports] where body
+            | body
+
+  alias -> [namespace] as modid
+
+  namespace -> data
+             | type
+  ```
+    * When `as modid` is specified, the users can refer to the module within
+      the module itself using the provided qualifier.
+    * When `data as modid` is specified, the users can refer to the terms
+      in the module within the module itself using the provided qualifier.
+    * When `type as modid` is specified, the users can refer to the types
+      in the module within the module itself using the provided qualifier.
+
+* Deprecate the `'` syntax of `-XDataKinds`, reserving this syntax for Template
+  Haskell name quotation. Introduce a new warning, `-Wticked-promoted-constructors` which
+  warns of the usage of the syntax.
+
+* Deprecate the `''` syntax in Template Haskell. Introduce a new warning,
+  `-Wdouble-tick-template-haskell` which warns about the usage of the syntax.
+
+* Introduce a new warning, `-Wpun-bindings`, triggered by any name
```

Comment:
Which warning bin? And would this warn if the clash is across different modules? Example:

```hs
module A where
data S = T
```

```hs
module B where
import A
data T = MkT
```

Does `data T` get the warning?

Reply:
Yes, it should trigger `-Wpun-bindings` because if Haskell had a single unified namespace there would be a name clash. I'll add this to examples.

Reply:
Take this example:

```haskell
module A where

foo :: String
foo = "foo"
```

```haskell
module B where

import A

foo :: String
foo = "bar"
```

This is a name clash, but it does not produce any error or warning. This seems to indicate that you shouldn't get a pun warning in @goldfirere's example.

Reply:
Oh, you are correct! I remember now, I tested this before but with types:
```
module A where
data T
```
```
module B where
import A
data T
```

This doesn't produce any errors until you try to use `T` (at which point you'll get ambiguous occurrence error). Completely forgot about this, since it's been a while since then!

Reply:
We might have some sort of "shadows imports" warning that is orthogonal to this.

Reply:
Interestingly enough, what happens now is that you get a `-Wname-shadowing` warning, but *only* if the binding on module `B` is a local binding, not if it's a top-level binding.